lnwallet/rpcwallet/remote_signer.go

```diff
@@ -0,0 +1,88 @@
+)
+
+// RemoteSigner is an interface that abstracts the communication with a remote
+// signer. It extends the RemoteSignerRequests, and adds some additional methods
+// to manage the connection and verify the health of the remote signer.
+type RemoteSigner interface {
+	// RemoteSignerRequests is an interface that defines the requests that
+	// can be sent to a remote signer.
+	RemoteSignerRequests
+
+	// Timeout returns the set connection timeout for the remote signer.
+	Timeout() time.Duration
+
+	// Ready blocks and returns nil when the remote signer is ready to
+	// accept requests.
+	Ready() error
+
+	// Ping verifies that the remote signer is still responsive.
+	Ping(timeout time.Duration) error
+
+	// Run feeds lnd with the incoming stream set up by an outbound remote
+	// signer and then blocks until the stream is closed. Lnd can then send
+	// any requests to the remote signer through the stream.
+	Run(stream walletrpc.WalletKit_SignCoordinatorStreamsServer) error
+}
+
+// RemoteSignerRequests is an interface that defines the requests that can be
+// sent to a remote signer. It's a subset of the signrpc.SignerClient and
+// walletrpc.WalletKitClient interfaces.
+type RemoteSignerRequests interface {
+	// DeriveSharedKey sends a SharedKeyRequest to the remote signer and
+	// waits for the corresponding response.
+	DeriveSharedKey(ctx context.Context,
+		in *signrpc.SharedKeyRequest,
+		opts ...grpc.CallOption) (*signrpc.SharedKeyResponse, error)
+
+	// DeriveSharedKey sends a MuSig2CleanupRequest to the remote signer and
+	// waits for the corresponding response.
+	MuSig2Cleanup(ctx context.Context,
+		in *signrpc.MuSig2CleanupRequest,
+		opts ...grpc.CallOption) (*signrpc.MuSig2CleanupResponse, error)
+
+	// MuSig2CombineSig sends a MuSig2CombineSigRequest to the remote signer
+	// and waits for the corresponding response.
+	MuSig2CombineSig(ctx context.Context,
+		in *signrpc.MuSig2CombineSigRequest,
+		opts ...grpc.CallOption) (*signrpc.MuSig2CombineSigResponse,
+		error)
+
+	// MuSig2CreateSession sends a MuSig2SessionRequest to the remote signer
+	// and waits for the corresponding response.
+	MuSig2CreateSession(ctx context.Context,
+		in *signrpc.MuSig2SessionRequest,
+		opts ...grpc.CallOption) (*signrpc.MuSig2SessionResponse, error)
+
+	// MuSig2RegisterNonces sends a MuSig2RegisterNoncesRequest to the
+	// remote signer and waits for the corresponding response.
+	MuSig2RegisterNonces(ctx context.Context,
+		in *signrpc.MuSig2RegisterNoncesRequest,
+		opts ...grpc.CallOption) (*signrpc.MuSig2RegisterNoncesResponse,
+		error)
+
+	// MuSig2Sign sends a MuSig2SignRequest to the remote signer and waits
+	// for the corresponding response.
+	MuSig2Sign(ctx context.Context,
+		in *signrpc.MuSig2SignRequest,
+		opts ...grpc.CallOption) (*signrpc.MuSig2SignResponse, error)
+
+	// SignMessage sends a SignMessageReq to the remote signer and waits for
+	// the corresponding response.
+	SignMessage(ctx context.Context,
```

Comment:
Oh wow, is this really the full list of things we need from a remote signer? Seems quite manageable, which will probably make our lives easier for future VLS integrations...

Reply:
IIUC this is part of what's needed to make a more loosely coupled remote signer implementation, so one that allows the "node" part to start up without the signer connection already established.

Reply:
Yes that's correct! These are all the operations the remote signer acts through currently, no matter if it's an inbound remote signer, or this new outbound remote signer version.

> Seems quite manageable, which will probably make our lives easier for future VLS integrations.

I'm not confident that this will be enough for VLS though unfortunately... VLS needs metadata in regards to all channel opens, closures and HTLCs etc. that occur on the node, so I'm not sure if we can properly just attach that metadata to these calls in a functional way, or if we'll need to specify individual methods which separate the calls more clearly. But it would be worth exploring if we can just attach the necessary metadata to these calls!